pubsub/iterator.go

```diff
@@ -639,35 +643,33 @@ func (it *messageIterator) sendAck(m map[string]*AckResult) {
 				s, ok := it.activeSpans.LoadAndDelete(ackID)
 				if ok {
 					subscribeSpan := s.(trace.Span)
-					if subscribeSpan.SpanContext().IsSampled() {
-						defer subscribeSpan.End()
-						defer subscribeSpan.SetAttributes(attribute.String(resultAttribute, resultAcked))
-						subscribeSpans = append(subscribeSpans, subscribeSpan)
-						subscribeSpan.AddEvent(eventAckStart, trace.WithAttributes(semconv.MessagingBatchMessageCount(len(ackIDs))))
-						defer subscribeSpan.AddEvent(eventAckEnd)
-						links = append(links, trace.Link{SpanContext: subscribeSpan.SpanContext()})
-					}
-				} else {
-					log.Printf("pubsub: sendAck failed to load ackID(%s) from activeSpans map", ackID)
+					defer subscribeSpan.End()
+					defer subscribeSpan.SetAttributes(attribute.String(resultAttribute, resultAcked))
+					subscribeSpans = append(subscribeSpans, subscribeSpan)
+					subscribeSpan.AddEvent(eventAckStart, trace.WithAttributes(semconv.MessagingBatchMessageCount(len(ackIDs))))
+					defer subscribeSpan.AddEvent(eventAckEnd)
+					links = append(links, trace.Link{SpanContext: subscribeSpan.SpanContext()})
 				}
 			}
 		}
 		var ackSpan trace.Span
 		if it.enableTracing {
-			opts := getCommonOptions(it.subID)
+			opts := getCommonOptions(it.projectID, it.subID)
 			opts = append(opts, trace.WithLinks(links...))
 			ctx, ackSpan = startSpan(context.Background(), ackSpanName, it.subID, opts...)
 			defer ackSpan.End()
 			ackSpan.SetAttributes(semconv.MessagingBatchMessageCount(len(ackIDs)),
 				semconv.CodeFunction("sendAck"))
 
-			for _, s := range subscribeSpans {
-				s.AddLink(trace.Link{
-					SpanContext: ackSpan.SpanContext(),
-					Attributes: []attribute.KeyValue{
-						semconv.MessagingOperationName(ackSpanName),
-					},
-				})
+			if ackSpan.SpanContext().IsSampled() {
```

Comment:
Just curious what the reasoning was for checking this. Is there an issue with linking to a span that isn't sampled? Not that this is a lot of code either way.

Reply:
In Cloud Trace, the link icon appears on the span, but it links to a NOT_FOUND page, which is poor UX in my opinion. We should only create links to spans that are actually sampled.